src/lib/operators/join_hash.cpp

```diff
@@ -217,69 +217,43 @@ std::shared_ptr<Partition<T>> materialize_input(const std::shared_ptr<const Tabl
-            // We need to add this to avoid gaps in the list of offsets when we iterate later on
-            materialized_chunk.emplace_back(NULL_ROW_ID, T{});
-          }
-        });
-      });
+            const uint32_t hashed_value = hash_value<T, HashedType>(value.value(), partitioning_seed);
+
+            /*
+            For ReferenceColumns we do not use the RowIDs from the referenced tables.
+            Instead, we use the index in the ReferenceColumn itself. This way we can later correctly dereference
+            values from different inputs (important for Multi Joins).
+            */
+            if constexpr (std::is_same<std::decay<decltype(typed_column)>, ReferenceColumn>::value) {
+              *(output_iterator++) =
+                  PartitionedElement<T>{RowID{chunk_id, reference_column_offset}, hashed_value, value.value()};
+            } else {
+              *(output_iterator++) =
+                  PartitionedElement<T>{RowID{chunk_id, value.chunk_offset()}, hashed_value, value.value()};
+            }
 
-      size_t row_id = output_offset;
-
-      /*
-      For ReferenceColumns we do not use the RowIDs from the referenced tables.
-      Instead, we use the index in the ReferenceColumn itself. This way we can later correctly dereference
-      values from different inputs (important for Multi Joins).
-      For performance reasons this if statement is around the for loop.
-      */
-      if (auto ref_column = std::dynamic_pointer_cast<const ReferenceColumn>(column)) {
-        // hash and add to the other elements
-        ChunkOffset offset = 0;
-        for (auto&& elem : materialized_chunk) {
-          if (elem.first.chunk_offset != INVALID_CHUNK_OFFSET) {
-            uint32_t hashed_value = hash_value<T, HashedType>(elem.second, partitioning_seed);
-            output[row_id] = PartitionedElement<T>{RowID{chunk_id, offset}, hashed_value, elem.second};
-
-            const Hash radix = (output[row_id].partition_hash >> (32 - radix_bits * (pass + 1))) & mask;
+            const Hash radix = (hashed_value >> (32 - radix_bits * (pass + 1))) & mask;
             histogram[radix]++;
-
-            row_id++;
           }
-
-          offset++;
-        }
-      } else {
-        // hash and add to the other elements
-        for (auto&& elem : materialized_chunk) {
-          if (elem.first.chunk_offset == INVALID_CHUNK_OFFSET) continue;
-
-          uint32_t hashed_value = hash_value<T, HashedType>(elem.second, partitioning_seed);
-          output[row_id] = PartitionedElement<T>{elem.first, hashed_value, elem.second};
-
-          const Hash radix = (output[row_id].partition_hash >> (32 - radix_bits * (pass + 1))) & mask;
-          histogram[radix]++;
-
-          row_id++;
-        }
-      }
+          // reference_column_offset is only used for ReferenceColumns
+          if constexpr (std::is_same<std::decay<decltype(typed_column)>, ReferenceColumn>::value) {
+            reference_column_offset++;
```

Comment:
I would have assumed that the compiler finds this itself. It's better to be explicit anyway, but did you test it?

Reply:
That's true. No, I didn't test it.

Maybe it would be easier to read without the if?

Reply:
All fine, I was just curious.

Reply:
Ok 😄